#165344089 Implement clapping functionality

devdbrandy opened the pull request:
#### What does this PR do?

This provides support for adding claps to articles

#### Description of Task to be completed?

- [x] Add migration for article claps
- [x] Add `ArticleClaps` model
- [x] Add endpoint for creating/update an article claps
- [x] Write test against implementation
- [x] Add validation for allowed number of claps

#### How should this be manually tested?

1. Run the following commands on terminal:

```bash
git clone -b feature/165344089/clapping-functionality --single-branch \
https://github.com/andela/apollo-ah-backend.git

cd apollo-ah-backend
cp .env.example .env

# To run test
cp .env.example .env.test

# Install package modules
npm install
```

2. Modify `.env` file with your db credentials

3. Run database migrations with `npm run migrate`

4. Run seeders with `npm run seed`

5. Spin up the server by running `npm run start:dev` on your terminal

6. Test new endpoints listed below 👇with [Postman](https://www.getpostman.com/downloads/).
- `POST /articles/{slug}/claps` - Add/update claps to specific article
   Request body - `claps` number of claps
- `GET /articles/{slug}/claps` - Get article claps

#### Running test locally

Two npm scripts are available to run the test suite:

- `npm run mocha` or `npm run mocha:watch` - The later watches for any file changes and runs the full test suite (without code coverage)
- `npm test` - Performs a single full test suite run, including Istanbul code coverage reporting. Summary coverage reports are written to stdout, and detailed HTML reports are available in /coverage/index.html

**NOTE:** To run test locally, make sure to create `.env.test` file.

#### Any background context you want to provide?

n/a

#### What are the relevant pivotal tracker stories?

[#165344089](https://www.pivotaltracker.com/story/show/165344089)

#### Screenshots (if appropriate)

n/a

devdbrandy committed:
feat(articleClaps): :sparkles: add clapping functionality

- Added migration for article claps
- Added `ArticleClaps` model
- Added endpoint for creating/update an article claps
- Wrote test against implementation
- Added validation for allowed number of claps

[#165344089]

valenstical commented on server/controllers/clapsController.js:
```diff
@@ -0,0 +1,123 @@
+    } = request;
+    let statusCode = STATUS.CREATED;
+    try {
+      const article = await models.Article.findOne({ where: { slug } });
```
~~Do you think we can move this validation to a middleware?~~

valenstical commented on test/integrations/routes/claps.test.js:
```diff
@@ -0,0 +1,117 @@
+      chai
+        .request(app)
+        .get(`/api/v1/articles/${articleSlug}/claps`)
+        .set({ Authorization: `Bearer ${userToken}` })
```
_Hmmmm_, this means, I must be logged in to view claps for a specific article. 
 Do you think it would also be OK for unauthorized users to view the number of claps for a specific article? Since this is not really a secret. 😊

devdbrandy replied:
Thats a good point @valenstical.
However, the idea behind this is to have an interface where both authenticated users and authors can get a list of claps, and by who. Regarding the number of claps, we already have that included as part of the result obtained when fetching article resource(s).

valenstical commented on test/integrations/routes/claps.test.js:
```diff
@@ -0,0 +1,117 @@
+          done();
+        });
+    });
+    it('Should throw validation error when claps exceeds 100', (done) => {
```
Can we also add validation test for when the clap parameter is not supplied or it's not a number?

devdbrandy replied:
Thanks for pointing that, I've implemented the requested changes.

valenstical commented on server/middlewares/validator.js:
```diff
@@ -338,4 +343,18 @@ export default class Validator {
+    return [
+      body('claps')
+        .isInt({ max: CLAPS_LIMIT })
+        .withMessage('Claps must not exceed 100'),
```
Would you consider adding a different validation to check if the clap parameter is an integer (or it's empty as well). 
With the current implementation, users will get the message '_Claps must not exceed 100_' if they,

- enter an invalid number
- or do not supply a clap parameter

valenstical reviewed (changes requested):
This is really good work @devdbrandy, we now have a clapping functionality. 
I have just a few requested changes. Good work once more.

devdbrandy committed:
feat(articleClap): modify controller to a more descriptive name

aanchirinah reviewed (approved):
LGTM

valenstical committed:
chore(claps): enhance clapping feature

- add validation to check if clap parameter is a valid integer
- add route GET articles/:slug/claps/:userId to fetch number of claps by a specific user
- write integration tests

valenstical reviewed (approved):
LGTM

devdbrandy committed:
feat(articleClaps): :zap: enhance clap functionality

- Improve logic
- Add article validation upon request
- Restrict users from clapping owned article
- Write addtional test against implementation

[#165344089]

devdbrandy committed:
:twisted_rightwards_arrows: resolve merge conflicts

devdbrandy committed:
fix(test): :ambulance: remove `only` syntax

devdbrandy committed:
fix(test): :green_heart: resolve travisCI build failure

devdbrandy committed:
feat(articleClaps): :egg: add default for user without claps

Default claps value to 0 for user without claps

devdbrandy committed:
Merge branch 'staging' into feature/165344089/clapping-functionality
